controllers/k8ssandra/add_dc_test.go

```diff
@@ -0,0 +1,522 @@
+package k8ssandra
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	cassctlapi "github.com/k8ssandra/cass-operator/apis/control/v1alpha1"
+	api "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	reaperapi "github.com/k8ssandra/k8ssandra-operator/apis/reaper/v1alpha1"
+	stargateapi "github.com/k8ssandra/k8ssandra-operator/apis/stargate/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/annotations"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/stargate"
+	testutils "github.com/k8ssandra/k8ssandra-operator/pkg/test"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	"github.com/k8ssandra/k8ssandra-operator/test/framework"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"testing"
+	"time"
+)
+
+// addDc tests scenarios that involve adding a new CassandraDatacenter to an existing
+// K8ssandraCluster.
+func addDc(t *testing.T, ctx context.Context, f *framework.Framework, namespace string) {
+	t.Run("WithUserKeyspaces", addDcTest(ctx, f, withUserKeyspaces))
+	t.Run("WithStargateAndReaper", addDcTest(ctx, f, withStargateAndReaper))
+	t.Run("FailSystemKeyspaceUpdate", addDcTest(ctx, f, failSystemKeyspaceUpdate))
+	t.Run("FailUserKeyspaceUpdate", addDcTest(ctx, f, failUserKeyspaceUpdate))
+}
+
+type addDcTestFunc func(ctx context.Context, t *testing.T, f *framework.Framework, kc *api.K8ssandraCluster)
+
+func addDcSetup(ctx context.Context, t *testing.T, f *framework.Framework, namespace string) *api.K8ssandraCluster {
+	require := require.New(t)
+	kc := &api.K8ssandraCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "add-dc-test",
+		},
+		Spec: api.K8ssandraClusterSpec{
+			Cassandra: &api.CassandraClusterTemplate{
+				Cluster: "test",
+				Datacenters: []api.CassandraDatacenterTemplate{
+					{
+						Meta: api.EmbeddedObjectMeta{
+							Name: "dc1",
+						},
+						K8sContext:    k8sCtx0,
+						Size:          3,
+						ServerVersion: "4.0.1",
+						StorageConfig: &cassdcapi.StorageConfig{
+							CassandraDataVolumeClaimSpec: &corev1.PersistentVolumeClaimSpec{
+								StorageClassName: &defaultStorageClass,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	kcKey := client.ObjectKey{Namespace: kc.Namespace, Name: kc.Name}
+
+	createSuperuserSecret(ctx, t, f, kcKey, kc.Spec.Cassandra.Cluster)
+
+	createReplicatedSecret(ctx, t, f, kcKey, "cluster-0")
```

Comment:
When adding a new cluster, we don't actually have any replication rule updates. I don't see here if we actually verify that some secrets are even copied to new clustrs? It's not strictly necessary per-datacenter of course, since datacenter could be replicated to an existing cluster in another namespace, but I wonder if these kind of issues came to mind (at least in documentation)

Reply:
`verifyReplicatedSecretReconciled` is called in each test. It verifies that the `ReplicationDone` condition gets set.